File: v1/src/main/java/com/api/v1/builders/CarBuilder.java
package com.api.v1.builders;

import com.api.v1.domain.Car;
import com.api.v1.dtos.CarRegistrationRequestDto;
import jakarta.validation.Valid;

import java.time.ZonedDateTime;
import java.util.UUID;

public class CarBuilder {

    private final UUID id = UUID.randomUUID();
    private String make;
    private String model;
    private int productionYear;
    private String vin;
    private String plateNumber;
    private final String addedAt = ZonedDateTime.now().toString();

    private CarBuilder() {}

    public static CarBuilder create() {
        return new CarBuilder();
    }

    public CarBuilder fromDto(@Valid CarRegistrationRequestDto requestDto) {
        this.make = requestDto.make();
        this.model = requestDto.model();
        this.productionYear = requestDto.productionYear();
        this.vin = requestDto.vin();
        this.plateNumber = requestDto.plateNumber();
        return this;
    }

    public Car build() {
        return new Car(
                id,
                make,
                model,
                productionYear,
                vin,
                plateNumber,
                addedAt
        );
    }

}

File: v2/src/main/kotlin/com/api/v2/car/CarBuilder.kt
package com.api.v2.car

import java.time.ZonedDateTime
import java.util.*

class CarBuilder {

    private val id: UUID = UUID.randomUUID()
    private lateinit var vin: String
    private lateinit var model: String
    private lateinit var productionYear: String
    private lateinit var make: String
    private lateinit var plateNumber: String
    private val createdAt: String = ZonedDateTime.now().toString()

    companion object {
        fun create(): CarBuilder {
            return CarBuilder()
        }
    }

    fun fromDto(requestDto: CarRegistrationRequestDto): CarBuilder {
        vin = requestDto.vin
        model = requestDto.model
        productionYear = requestDto.productionYear
        make = requestDto.make
        plateNumber = requestDto.plateNumber
        return this
    }

    fun build(): Car {
        return Car(
            id,
            vin,
            model,
            productionYear,
            make,
            plateNumber,
            createdAt
        )
    }

}